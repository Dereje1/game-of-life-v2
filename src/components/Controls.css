.MuiSlider-root .MuiSlider-markLabel {
  font-size: 1vmax;
}
