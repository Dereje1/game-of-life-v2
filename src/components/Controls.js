import React from 'react';
import IconButton from '@mui/material/IconButton';
import PlayCircleFilledWhiteIcon from '@mui/icons-material/PlayCircleFilledWhite';
import PauseIcon from '@mui/icons-material/Pause';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import HighlightOffOutlinedIcon from '@mui/icons-material/HighlightOffOutlined';
import Slider from '@mui/material/Slider';
import Chip from '@mui/material/Chip';
import './Controls.css'

export const Action = ({ isRefreshing }) => {
    if (!isRefreshing) {
        return <PlayCircleFilledWhiteIcon style={{ fontSize: '2em', color: 'green' }} />
    } else {
        return <PauseIcon style={{ fontSize: '2em', color: 'red' }} />
    }
}

export const ControlTop = ({
    width,
    height,
    onRefresh,
    onPause,
    isRefreshing,
    onReset,
    onClear,
    generations
}) => (
    <div
        style={{
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-around',
            width,
            height
        }}
    >
        <div>
            <IconButton aria-label="play_pause" onClick={isRefreshing ? onPause : onRefresh}>
                <Action isRefreshing={isRefreshing} />
            </IconButton>
            <IconButton aria-label="reset" onClick={onReset}>
                <RestartAltIcon style={{ fontSize: '2em', color: 'blue' }} />
            </IconButton>
            <IconButton aria-label="clear" onClick={onClear}>
                <HighlightOffOutlinedIcon style={{ fontSize: '2em' }} />
            </IconButton>
        </div>
        <div>
            <Chip label={generations} style={{ width: 100 }} color={"success"} onClick={() => window.open('https://pi.math.cornell.edu/~lipa/mec/lesson6.html')} />
        </div>
    </div>
)

export const ControlBottom = ({
    width,
    height,
    handleCellSize,
    cellSize,
    onRefreshRate,
    refreshRate
}) => (
    <div
        style={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'space-around',
            width,
            height
        }}
    >
        <div style={{ width: width / 3 }}>
            <span style={{margin: 0, fontSize: '.75em', fontWeight: 'bold'}}>Cell Size</span>
            <Slider
                aria-label="Granularity"
                defaultValue={20}
                valueLabelDisplay="off"
                step={5}
                marks={[
                    {
                        value: 10,
                        label: '10',
                    },
                    {
                        value: 15,
                        label: '15',
                    },
                    {
                        value: 20,
                        label: '20',
                    },
                    {
                        value: 25,
                        label: '25',
                    },
                    {
                        value: 30,
                        label: '30px',
                    }
                ]}
                min={10}
                max={30}
                onChange={handleCellSize}
                value={cellSize}
            />
        </div>
        <div style={{ width: width / 3 }}>
            <span style={{margin: 0, fontSize: '.75em', fontWeight: 'bold'}}>Referesh Rate</span>
            <Slider
                aria-label="Refreshrate"
                defaultValue={120}
                valueLabelDisplay="off"
                step={125}
                marks={[
                    {
                        value: 20,
                        label: '20',
                    },
                    {
                        value: 140,
                        label: '140',
                    },
                    {
                        value: 260,
                        label: '260',
                    },
                    {
                        value: 380,
                        label: '380',
                    },
                    {
                        value: 500,
                        label: '500ms',
                    }
                ]}
                min={20}
                max={500}
                onChange={onRefreshRate}
                value={refreshRate}
                color="secondary"
            />
        </div>
    </div>
)
