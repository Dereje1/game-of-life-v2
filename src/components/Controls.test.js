import React from 'react'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ControlBottom, ControlTop, Action, Links } from './Controls'

const controlTopProps = {
    width: 80,
    height: 80,
    handleRefresh: jest.fn(),
    handlePause: jest.fn(),
    isRefreshing: true,
    handleReset: jest.fn(),
    handleClear: jest.fn(),
    generations: 6
}

const controlBottomProps = {
    width: 80,
    height: 80,
    handleCellSize: jest.fn(),
    handleRefreshRate: jest.fn(),
    cellSize: 20,
    refreshRate: 140,
}

test('will render the top control', () => { 
    const wrapper = shallow(<ControlTop {...controlTopProps}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('will render the bottom control', () => { 
    const wrapper = shallow(<ControlBottom {...controlBottomProps}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('will render play when not refreshing', () => { 
    const wrapper = shallow(<Action isRefreshing={false}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('will render pause when refreshing', () => { 
    const wrapper = shallow(<Action isRefreshing={true}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
})

test('will render links', () => { 
    const wrapper = shallow(<Links />)
    expect(toJson(wrapper)).toMatchSnapshot();
})