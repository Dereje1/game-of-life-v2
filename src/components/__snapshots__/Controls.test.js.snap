// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render pause when refreshing 1`] = `
<Memo(ForwardRef(PauseIcon))
  style={
    Object {
      "color": "#d32f2f",
      "fontSize": "2.5rem",
    }
  }
/>
`;

exports[`will render play when not refreshing 1`] = `
<Memo(ForwardRef(PlayCircleFilledWhiteIcon))
  style={
    Object {
      "color": "#0288d1",
      "fontSize": "2.5rem",
    }
  }
/>
`;

exports[`will render the bottom control 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": 80,
      "justifyContent": "space-around",
      "marginTop": 5,
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "width": "30%",
      }
    }
  >
    <ForwardRef(Typography)
      gutterBottom={true}
      id="linear-slider"
      sx={
        Object {
          "fontSize": "1.3vmax",
          "fontWeight": "bold",
          "marginBottom": 0,
        }
      }
    >
      Element size: 20px
    </ForwardRef(Typography)>
    <ForwardRef(Slider)
      aria-label="Cellsize"
      defaultValue={15}
      marks={
        Array [
          Object {
            "value": 5,
          },
          Object {
            "value": 10,
          },
          Object {
            "value": 15,
          },
          Object {
            "value": 20,
          },
          Object {
            "value": 25,
          },
        ]
      }
      max={25}
      min={5}
      onChange={[MockFunction]}
      step={5}
      value={20}
      valueLabelDisplay="off"
    />
  </div>
  <div
    style={
      Object {
        "width": "30%",
      }
    }
  >
    <ForwardRef(Typography)
      gutterBottom={true}
      id="non-linear-slider"
      sx={
        Object {
          "fontSize": "1.3vmax",
          "fontWeight": "bold",
          "marginBottom": 0,
        }
      }
    >
      8 generations / sec
    </ForwardRef(Typography)>
    <ForwardRef(Slider)
      aria-label="Refreshrate"
      color="secondary"
      defaultValue={3}
      marks={
        Array [
          Object {
            "value": 1,
          },
          Object {
            "value": 2,
          },
          Object {
            "value": 3,
          },
          Object {
            "value": 6,
          },
        ]
      }
      max={6}
      min={1}
      onChange={[MockFunction]}
      scale={[Function]}
      step={null}
      value={3}
      valueLabelDisplay="off"
    />
  </div>
</div>
`;

exports[`will render the top control 1`] = `
<div
  style={
    Object {
      "background": "#e5e5e359",
      "display": "flex",
      "flexDirection": "column",
      "height": 80,
      "justifyContent": "flex-end",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "width": "100%",
      }
    }
  >
    <div>
      <ForwardRef(IconButton)
        aria-label="play_pause"
        disabled={false}
        onClick={[MockFunction]}
      >
        <Action
          disabled={false}
          isRefreshing={true}
        />
      </ForwardRef(IconButton)>
      <ForwardRef(IconButton)
        aria-label="next_step"
        disabled={true}
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(SkipNextIcon))
          style={
            Object {
              "fontSize": "2.5rem",
            }
          }
        />
      </ForwardRef(IconButton)>
      <ForwardRef(IconButton)
        aria-label="clear"
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(HighlightOffOutlinedIcon))
          style={
            Object {
              "fontSize": "2.5rem",
            }
          }
        />
      </ForwardRef(IconButton)>
      <ForwardRef(IconButton)
        aria-label="settings"
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(SettingsIcon))
          color="error"
          style={
            Object {
              "fontSize": "2rem",
            }
          }
        />
      </ForwardRef(IconButton)>
    </div>
    <div>
      <ForwardRef(Chip)
        color="error"
        label={6}
        style={
          Object {
            "justifyContent": "space-between",
            "marginBottom": 3,
            "width": 100,
          }
        }
        variant="outlined"
      />
    </div>
  </div>
  <ForwardRef(Tooltip)
    placement="top-start"
    title="7.0 generations/sec"
  >
    <ForwardRef(LinearProgress)
      sx={
        Object {
          "& .MuiLinearProgress-bar": Object {
            "bgcolor": "#d32f2f",
          },
          "background": "#e5e5e359",
        }
      }
      value={87}
      variant="determinate"
    />
  </ForwardRef(Tooltip)>
</div>
`;
